### ISSUE TITLE
PKCS7_verify can add errors to stack even when successful - alpha17

### ISSUE THREAD

**Issue by reaperhulk** [openssl/openssl#15389]
Similar to #14996 it is possible to make a call to `PKCS7_verify` that succeeds in the 3.0.0 alphas, but adds errors to the stack. Reproducer:

```c
#include <openssl/bio.h>
#include <openssl/pkcs7.h>
#include <openssl/pem.h>
#include <openssl/err.h>
#include <assert.h>
#include <string.h>

int main(void) {
    int res = 0;
    const char *sig = "MIME-Version: 1.0\nContent-Type: multipart/signed; protocol=\"application/x-pkcs7-signature\"; micalg=\"sha-256\"; boundary=\"----9B5319FF2E4428B17CD26B69294E7F31\"\n\nThis is an S/MIME signed message\n\n------9B5319FF2E4428B17CD26B69294E7F31\nContent-Type: text/plain\r\n\r\nhello world\n------9B5319FF2E4428B17CD26B69294E7F31\nContent-Type: application/x-pkcs7-signature; name=\"smime.p7s\"\nContent-Transfer-Encoding: base64\nContent-Disposition: attachment; filename=\"smime.p7s\"\n\nMIIDEgYJKoZIhvcNAQcCoIIDAzCCAv8CAQExDzANBglghkgBZQMEAgEFADALBgkq\nhkiG9w0BBwGgggFVMIIBUTCB96ADAgECAgIDCTAKBggqhkjOPQQDAjAnMQswCQYD\nVQQGEwJVUzEYMBYGA1UEAwwPY3J5cHRvZ3JhcGh5IENBMB4XDTE3MDEwMTEyMDEw\nMFoXDTM4MTIzMTA4MzAwMFowJzELMAkGA1UEBhMCVVMxGDAWBgNVBAMMD2NyeXB0\nb2dyYXBoeSBDQTBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABBj/z7v5Obj13cPu\nwECLBnUGq0/N2CxSJE4f4BBGZ7VfFblivTvPDG++Gve0oQ+0uctuhrNQ+WxRv8GC\n177F+QWjEzARMA8GA1UdEwEB/wQFMAMBAf8wCgYIKoZIzj0EAwIDSQAwRgIhANES\n742XWm64tkGnz8DnpG6u2lHkZFQr3oaVvPcemvlbAiEA0WGGzmYx5C9UvfXIK7NE\nziT4pQtyESE0uRVKXw4nMqkxggGBMIIBfQIBATAtMCcxCzAJBgNVBAYTAlVTMRgw\nFgYDVQQDDA9jcnlwdG9ncmFwaHkgQ0ECAgMJMA0GCWCGSAFlAwQCAQUAoIHkMBgG\nCSqGSIb3DQEJAzELBgkqhkiG9w0BBwEwHAYJKoZIhvcNAQkFMQ8XDTIxMDUyMDE4\nNTA0OVowLwYJKoZIhvcNAQkEMSIEIOdwMRgQrqcnmMYvag+BVvErcc6bwUXI94Ds\nQkiyIU9pMHkGCSqGSIb3DQEJDzFsMGowCwYJYIZIAWUDBAEqMAsGCWCGSAFlAwQB\nFjALBglghkgBZQMEAQIwCgYIKoZIhvcNAwcwDgYIKoZIhvcNAwICAgCAMA0GCCqG\nSIb3DQMCAgFAMAcGBSsOAwIHMA0GCCqGSIb3DQMCAgEoMAoGCCqGSM49BAMCBEcw\nRQIhANYMJku1fW9T1MIEcAyREArz9kXCY4tWck5Pt0xzrYhaAiBDSP6e43zj4YtI\nuvQW+Lzv+dNF8EPuhgoPNe17RuUSLw==\n\n------9B5319FF2E4428B17CD26B69294E7F31--\n\n";
    const char *signed_data = "Content-Type: text/plain\r\n\r\nhello world";
    BIO *msg_bio = BIO_new_mem_buf(signed_data, strlen(signed_data));
    const char *cert_pem = "-----BEGIN CERTIFICATE-----\nMIIBUTCB96ADAgECAgIDCTAKBggqhkjOPQQDAjAnMQswCQYDVQQGEwJVUzEYMBYG\nA1UEAwwPY3J5cHRvZ3JhcGh5IENBMB4XDTE3MDEwMTEyMDEwMFoXDTM4MTIzMTA4\nMzAwMFowJzELMAkGA1UEBhMCVVMxGDAWBgNVBAMMD2NyeXB0b2dyYXBoeSBDQTBZ\nMBMGByqGSM49AgEGCCqGSM49AwEHA0IABBj/z7v5Obj13cPuwECLBnUGq0/N2CxS\nJE4f4BBGZ7VfFblivTvPDG++Gve0oQ+0uctuhrNQ+WxRv8GC177F+QWjEzARMA8G\nA1UdEwEB/wQFMAMBAf8wCgYIKoZIzj0EAwIDSQAwRgIhANES742XWm64tkGnz8Dn\npG6u2lHkZFQr3oaVvPcemvlbAiEA0WGGzmYx5C9UvfXIK7NEziT4pQtyESE0uRVK\nXw4nMqk=\n-----END CERTIFICATE-----";
    BIO *x509_bio = BIO_new_mem_buf(cert_pem, strlen(cert_pem));
    X509 *cert = PEM_read_bio_X509(x509_bio, NULL, NULL, NULL);
    assert(cert != NULL);
    assert(ERR_peek_error() == 0);

    X509_STORE *store = X509_STORE_new();
    assert(store != NULL);
    res = X509_STORE_add_cert(store, cert);
    assert(res == 1);

    BIO *bio = BIO_new_mem_buf(sig, strlen(sig));
    PKCS7 *p7 = SMIME_read_PKCS7(bio, NULL);
    assert(p7 != NULL);
    assert(ERR_peek_error() == 0);
    int flags = PKCS7_TEXT;
    res = PKCS7_verify(p7, NULL, store, msg_bio, NULL, flags);
    assert(res == 1);
    if (ERR_peek_error() != 0) {
        printf("error on stack but verify succeeded\n");
        return 1;
    }
}
```

**Comment by slontis** [comment]
Thanks for the test. It would of taken quite a while to find this without your very specific test case..

**Comment by slontis** [comment]
Back in mid January the following commit 
https://github.com/openssl/openssl/commit/dc88a0390608ccba76df6a26358215d1c0592948#diff-310dbb5e976563a48efb10737adf53420287a20eed4d1d1b0750450f38bf6809
subtly changed the behavior of BIO_write to handle NULL differently..

i.e.
````
bio_write_intern(BIO * b, …) changed from 

   If (b == NULL)
       return 0;

TO

    if (b == NULL) {
        ERR_raise(ERR_LIB_BIO, BIO_R_NULL_PARAMETER);
        return -1;
    }
````
There are quite a few functions that print and do not check the error code.
This potentially puts an error on the error stack and returns no error. This in not ideal.

So the options are:

1) Revert the change - but maybe this was done to fix some other issue?
2) Fix up all the places where BIO_write is not checked..
    For diagnotistic or printing functions should they just do:
      if (bio == NULL)
          return 1;  /* ignore */

@openssl/otc and @ddvo - does anyone have strong opinions on this?

**Comment by t8m** [comment]
I'd just revert that part of the dc88a03

**Comment by t8m** [comment]
It could affect 3rd party applications as well.

**Comment by richsalz** [comment]
It's a silent change to existing semantics that will most likely affect other users as @t8m points out.  Revert it.

**Comment by DDvO** [comment]
> Back in mid January the following commit commit dc88a0390608ccba76df6a26358215d1c0592948
> subtly changed the behavior of BIO_write to handle NULL differently..
> 
> i.e.
> 
> ```
> bio_write_intern(BIO * b, …) changed from 
> 
>    If (b == NULL)
>        return 0;
> 
> TO
> 
>     if (b == NULL) {
>         ERR_raise(ERR_LIB_BIO, BIO_R_NULL_PARAMETER);
>         return -1;
>     }
> ```
> 
> There are quite a few functions that print and do not check the error code.
> This potentially puts an error on the error stack and returns no error. This in not ideal.
> 
> So the options are:
> 
>     1. Revert the change - but maybe this was done to fix some other issue?
> 
>     2. Fix up all the places where BIO_write is not checked..
>        For diagnotistic or printing functions should they just do:
>        if (bio == NULL)
>        return 1;  /* ignore otherwise we break how it behaved before */
> 
> 
> @openssl/otc and @DDvO - does anyone have strong opinions on this?

Thanks for bringing this up.
As far as I recall, the error reporting I added there was not to fix a specific issue but just to follow the general model of doing input validation.
Yet I was not aware that for `BIO_write_{,ex}()` and thus also for `BIO_write_intern()` it should be legitimate to call them with `bio == NULL`, also because this feature is not documented.
So I propose reverting the change quoted here and documenting the option of NULL BIO input to `BIO_write_{,ex}()`.

**Comment by h-vetinari** [comment]
#15493 says "Fixes #15389", but the merged version https://github.com/openssl/openssl/commit/f505161e62d558b3f8442b264ccbf8112ebd58ef omitted that line. I presume this issue can be closed.

**Comment by slontis** [comment]
Thanks @h-vetinari..
